crates/resolver/src/error.rs

```diff
@@ -139,59 +138,83 @@ impl ResolveError {
                     query: Box::new(query),
                     soa: soa.map(Box::new),
                     negative_ttl: None,
-                    response_code: ResponseCode::ServFail,
+                    response_code,
                     trusted: false,
                 };
 
                 Err(ResolveError::from(error_kind))
             }
             // Some NXDOMAIN responses contain CNAME referals, that will not be an error
-            ResponseCode::NXDomain if !response.contains_answer() => {
-                let note = "Nameserver responded with NXDomain";
-                debug!("{}", note);
-
-                // TODO: if authoritative, this is cacheable, store a TTL (currently that requires time, need a "now" here)
-                // let valid_until = if response.is_authoritative() { now + response.get_negative_ttl() };
-
-                let mut response = response;
-                let soa = response.soa();
-                let negative_ttl = response.negative_ttl();
-
-                let query = response.take_queries().drain(..).next().unwrap_or_default();
-                let error_kind = ResolveErrorKind::NoRecordsFound {
-                    query: Box::new(query),
-                    soa: soa.map(Box::new),
-                    negative_ttl,
-                    response_code: ResponseCode::NXDomain,
-                    trusted: trust_nx,
-                };
-
-                Err(ResolveError::from(error_kind))
+            ResponseCode::NXDomain => {
+                if !response.contains_answer() {
```

Comment:
should the positive case be first? it's also a lot shorter.

Reply:
obsolete; I moved the positive case back out of the match arms, and added unqualified `ResponseCode::NXDomain | ResponseCode::NoError` arms below to cover the positive case.